### ISSUE TITLE
Message Board App

### ISSUE THREAD

**Issue by FFX01** [Cabal-Syndicate/randora_website#6]
This issue is for the discussion of the design and implementation of the Randora message board.

**Scope**
This is going to be the most difficult part of the website. We will need to build and host a secure database of posts and users. It's very important that we focus on security here.

**Structure**
As I see it, we really have two options here. We can either use MongoDB or SQL. MongoDB is supposedly faster and easier to implement, but SQL has a well proven track record. We can format the message boards like a traditional forum, like Voat/Reddit, or like an image board(4chan). This is up for discussion.

**Navigation**
This really depends on what DB we want to build and what format we want to use. Regardless, everything should be searchable

Please share your thoughts.

**Comment by Master-Foo** [comment]
I agree about the security. I disagree about the general difficulty. We should be able to mold the Djano tutorial into a halfway decent message board. Let's take this a bite at a time though. We start with the completed tutorial and just assign small features to be added.

Security though, should probably have it's own issue. I'll start one as soon as I'm done here.

**Database Stuff**
Django's ORM is designed in a way that specific databases doesn't matter. The purpose of ORM is to abstract away the database such that a developer doesn't need to be aware of it. He just writes the fucking code and it works.

I'll decide later which database we will use. Most likely Postgres, or MySQL. For development purposes everyone should be using SQLite. We can migrate between databases easily. If anyone wants to learn more I will show them.

However, (I say this as politely as possible), **WE WILL NOT BE USING MONGODB** I will rarely put my foot down. But this is the few times I will do it. Please trust that I have enough personal experience here. I will debate the issue if you want to know my reasoning, but the decision **IS** made. No MongoDB, or Reddis.

@FFX01 I agree about **Navigation**

**Comment by Master-Foo** [comment]
@FFX01 I noticed you decided to own this issue. Good hustle!

Before you dive into it, let's get a game plan going. I'll spend some time writing up a spec.

**Comment by Master-Foo** [comment]
So, I think the best way to do this is to make the message board a generic support app for other apps.

So, for instance, a user would have a message board. A team would have a message board. A page in the docs would have a message board. A page in the manual would have a message board.

So there wouldn't be a message board for the sake of having a message board. It exists to support each app in the system. So you wouldn't visit the "Site's Message Board", You'd visit the "Team's Message Board", or the "RNDObject Message Board".

Or, maybe a better explanation is "Whereever you are on the site, there is a relevant messsage board on that page for that specific topic".

Does this make sense?

So, The best thing to do at this point is to make a couple of apps and then make the message board app.

We already have the User App so we can start there. Let's make the Teams App next. (I'll start a new issue for this). When we have at least those two Apps, then we can make the relations for the message board.

Let me know what you think.

**Comment by Master-Foo** [comment]
I just opened the `Teams` App issue. Please refer to that before we start on this issue.

**Comment by FFX01** [comment]
I've already started chiding the database models and urls, these won't
change regardless of implementation.

We can simply draw from these databases on each page, but this will
undermine the modular structure of the site. I think embedding these views
inside of the templates from the other modules is the best approach as it
allows us to be DRY and maintain separation between modules. This will
become more important as the site grows and we begin to implement unit
testing. It will also be easier to maintain in the long run.

**Comment by Master-Foo** [comment]
OK, be sure to checkout and work on a new branch, so we don't get muddled commits. So, for instance, make a "Message Board" branch. This way we aren't accidentally committing partially baked "Message Board" code when we are actually submitting the "Teams" App.

As for undermining the modularity of the site, this is fine. The nature of having dependencies is going to require pre-requisite modules.

The point of the Apps are two fold.
1. To encourage "Everything in it's place". (Which is what we want)
2. To re-distribute an app. (Which isn't our priority, we aren't making apps)

Just because we put common functionality in one "App" doesn't necessarily mean other apps can't also depend on it. Nor does it mean that app can't have dependencies on other apps.

Apps aren't strictly meant to exist as an island. They are meant to add functionality to other apps.

If further proof of this is needed, look into the "User", "Group", "Auth" trifecta. They are separate apps within the default code base. However, they are meant to work together, dependant upon each other.

As for the views dilemma, I have a solution for this to keep "Dry". But let's get basic functionality in place first.
